settings.org
#+title:  Emacs configuration file
#+author: Berat Ertural
#+date:   2024-01-03

* Custom Look and Feel 
** Personal Information
This is me!

#+BEGIN_SRC emacs-lisp
  (setq user-full-name "Berat Ertural"
        user-mail-address "beertural+git@gmail.com")
#+END_SRC

** Default Behavior
Lets have some sane defaults, shall we?

#+BEGIN_SRC emacs-lisp
  ;; Optimization
  (setq gc-cons-threshold 100000000 ;  100mb
        read-process-output-max (* 1024 1024)) ; 1 mb
  
  (defalias 'yes-or-no-p 'y-or-n-p)
  
  ;; Keep all backup and auto-save files in one directory
  (setq backup-directory-alist '(("." . "~/.emacs.d/backups")))
  (setq auto-save-file-name-transforms '((".*" "~/.emacs.d/auto-save-list/" t)))
  
  ;; UTF-8 encoding
  (set-language-environment "UTF-8")
  (setq locale-coding-system 'utf-8)
  (set-terminal-coding-system 'utf-8)
  (set-keyboard-coding-system 'utf-8)
  (set-selection-coding-system 'utf-8)
  (prefer-coding-system 'utf-8)
  
  (setq-default indent-tabs-mode nil
                indicate-empty-lines t
                tab-width 4)
  
  ;; Enable Minor Modes
  (ido-mode 1)
  (ido-everywhere)
  (setq ido-enable-flex-matching t)
  ;(setq inhibit-startup-screen t)
  (tool-bar-mode -1)
  (menu-bar-mode -1)
  (show-paren-mode t)
  (global-linum-mode t)
  (column-number-mode t)
  (delete-selection-mode t)
  (global-visual-line-mode)
  
  ;; Global Keys
  (global-set-key (kbd "C-x C-b") 'ibuffer)
  (global-set-key (kbd "C-c w") 'whitespace-mode)
  (global-set-key (kbd "RET") 'newline-and-indent)

#+END_SRC
  
** Style
*** Theme 
Looking smart

#+BEGIN_SRC emacs-lisp
  (add-to-list 'custom-theme-load-path "~/.emacs.d/themes/")

  ;(load-theme 'tango-dark t)
  (load-theme 'flatland t)
#+END_SRC

*** Font   
Smexy fonts ^3^
#+BEGIN_SRC emacs-lisp
  (add-to-list 'default-frame-alist '(font . "FiraCode-10"))
  (set-face-attribute 'default t :font "FiraCode-10")

  ;; Font Ligatures
  ; Enable the www ligature in every possible major mode
  (require 'ligature)
  (ligature-set-ligatures 't '("www"))

  ; Enable ligatures in programming modes                                                           
  (ligature-set-ligatures 'prog-mode '("www" "**" "***" "**/" "*>" "*/" "\\\\" "\\\\\\" "{-" "::"
                                       ":::" ":=" "!!" "!=" "!==" "-}" "----" "-->" "->" "->>"
                                       "-<" "-<<" "-~" "#{" "#[" "##" "###" "####" "#(" "#?" "#_"
                                       "#_(" ".-" ".=" ".." "..<" "..." "?=" "??" ";;" "/*" "/**"
                                       "/=" "/==" "/>" "//" "///" "&&" "||" "||=" "|=" "|>" "^=" "$>"
                                       "++" "+++" "+>" "=:=" "==" "===" "==>" "=>" "=>>" "<="
                                       "=<<" "=/=" ">-" ">=" ">=>" ">>" ">>-" ">>=" ">>>" "<*"
                                       "<*>" "<|" "<|>" "<$" "<$>" "<!--" "<-" "<--" "<->" "<+"
                                       "<+>" "<=" "<==" "<=>" "<=<" "<>" "<<" "<<-" "<<=" "<<<"
                                       "<~" "<~~" "</" "</>" "~@" "~-" "~>" "~~" "~~>" "%%"))

  ;(global-ligature-mode 't)
#+END_SRC

* Navigation and Help
** Which-key
Handy if you forget the specific keybindings. You only need to remember the mode-map key.

#+BEGIN_SRC emacs-lisp
  (use-package which-key
    :ensure t
    :config
    (which-key-mode))
#+END_SRC  

** Company

Auto completion and help

#+BEGIN_SRC emacs-lisp
  (use-package company
    :bind (:map company-active-map
           ("C-n" . company-select-next)
           ("C-p" . company-select-previous))
    :config
      (setq tab-always-indent 'complete
        company-idle-delay 0.3
        company-selection-wrap-around t
        company-tooltip-align-annotations t
        company-idle-delay 0.36
        company-minimum-prefix-length 2
        company-tooltip-limit 10)
      (global-company-mode t))

  (use-package company-quickhelp
    :ensure t
    :config
      (company-quickhelp-mode)
      (setq company-quickhelp-delay 0.3))


  (eval-after-load 'company
    '(define-key company-active-map (kbd "C-c h") #'company-quickhelp-manual-begin))
#+END_SRC  

** Flycheck

Syntax checking

#+BEGIN_SRC emacs-lisp
  (use-package flycheck
    :ensure t
    :init
    (global-flycheck-mode t))

#+END_SRC  

* Text Editing
** Custom 

#+BEGIN_SRC emacs-lisp

  (defun increment-number-at-point ()
    (interactive)
    (skip-chars-backward "0-9")
    (or (looking-at "[0-9]+")
        (error "No number at point"))
    (replace-match (number-to-string (1+ (string-to-number (match-string 0))))))

  (global-set-key (kbd "C-c +") 'increment-number-at-point)

#+END_SRC

** Outline
#+BEGIN_SRC emacs-lisp

  ; Outline-minor-mode key map
  (define-prefix-command 'cm-map nil "Outline-")
                                          ; HIDE
  (define-key cm-map "q" 'hide-sublevels); Hide everything but the top-level headings
  (define-key cm-map "t" 'hide-body)     ; Hide everything but headings (all body lines)
  (define-key cm-map "o" 'hide-other)    ; Hide other branches
  (define-key cm-map "c" 'hide-entry)    ; Hide this entry's body
  (define-key cm-map "l" 'hide-leaves)   ; Hide body lines in this entry and sub-entries
  (define-key cm-map "d" 'hide-subtree)  ; Hide everything in this entry and sub-entries
                                          ; SHOW
  (define-key cm-map "a" 'show-all)      ; Show (expand) everything
  (define-key cm-map "e" 'show-entry)    ; Show this heading's body
  (define-key cm-map "i" 'show-children) ; Show this heading's child sub-headings
  (define-key cm-map "k" 'show-branches) ; Show all sub-headings under this heading
  (define-key cm-map "s" 'show-subtree)  ; Show everything in this heading & below
                                          ; MOVE
  (define-key cm-map "u" 'outline-up-heading)                ; Up
  (define-key cm-map "n" 'outline-next-visible-heading)      ; Next
  (define-key cm-map "p" 'outline-previous-visible-heading)  ; Previous
  (define-key cm-map "f" 'outline-forward-same-level)        ; Forward - same level
  (define-key cm-map "b" 'outline-backward-same-level)       ; Backward - same level
  (global-set-key "\M-o" cm-map)

#+END_SRC  

** Org-mode

#+BEGIN_SRC emacs-lisp
  (add-hook 'org-mode-hook 'flyspell-mode)
  (setq org-startup-numerated t)
  (setq org-format-latex-options (plist-put org-format-latex-options :scale 1.5))

  (setq org-latex-pdf-process
        (let
            ((cmd (concat "pdflatex -shell-escape -interaction nonstopmode"
                          " --synctex=1"
                          " -output-directory %o %f")))
          (list cmd
                "cd %o; if test -r %b.idx; then makeindex %b.idx; fi"
                "cd %o; bibtex %b"
                cmd
                cmd)))
#+END_SRC
  
** LaTeX

#+BEGIN_SRC emacs-lisp
  (add-hook 'LaTeX-mode-hook 'flyspell-mode)
  (add-hook 'LaTeX-mode-hook 'LaTeX-math-mode)
  (add-hook 'LaTeX-mode-hook 'turn-on-reftex)
  (add-hook 'LaTeX-mode-hook #'outline-minor-mode)


  (setq-default TeX-master t)
  (setq-default TeX-auto-save t)
  (setq-default TeX-parse-self t)
  ;(setq-default TeX-show-compilation t)
  (setq-default reftex-plug-into-AUCTeX t)
  (setq-default TeX-PDF-mode t)

  (require 'openwith)
  (openwith-mode t)
  (setq openwith-associations '(("\\.pdf\\'" "evince" (file))))
#+END_SRC

** Markdown

#+BEGIN_SRC emacs-lisp
  (use-package markdown-mode
    :defer 1
    :ensure t
    :mode ("README\\.md\\'" . gfm-mode)
    :init (setq markdown-command "multimarkdown"))
#+END_SRC

** Flyspell

Check your grammar.

#+BEGIN_SRC emacs-lisp
  (add-hook 'text-mode-hook 'flyspell-mode)
  (autoload 'flyspell-mode "flyspell" "On-the-fly ispell." t)
  (setq flyspell-issue-welcome-flag nil)

  (defun ispell-set-deutsch ()
    "Switch ispell language to DAS DEUTSCHE SPRACHE!"
    (interactive)
    (ispell-change-dictionary "deutsch8"))

  (defun ispell-set-english ()
    "Switch ispell language to FREEDOM (⌐■_■)"
    (interactive)
    (ispell-change-dictionary "american"))

  (defvar toggle-ispell-deutsch t)
  (make-variable-buffer-local 'toggle-ispell-deutsch)

  (defun toggle-ispell-deutsch ()
    "Toggle ispell-language between english and deutsch8"
    (interactive)
    (cond 
      (toggle-ispell-deutsch
        (setq is-ispell-deutsch nil)
        (ispell-set-deutsch))
      (t
        (setq toggle-ispell-deutsch t)
        (ispell-set-english))))

  (add-hook 'post-mode-hook
    '(lambda ()
      (flyspell-mode t)
      (setq flyspell-generic-check-word-p 'mail-mode-flyspell-verify)))

#+END_SRC

** Treemacs
...
#+BEGIN_SRC emacs-lisp
  (add-hook 'emacs-startup-hook 'treemacs)
#+END_SRC

* Programming
** General

#+BEGIN_SRC emacs-lisp

  ;; Show unncessary whitespace that can mess up your diff
  (add-hook 'prog-mode-hook
            (lambda () (interactive) (setq show-trailing-whitespace 1)))

  ;; C/C++
  (add-hook 'c-mode-hook 'lsp)
  (add-hook 'c++-mode-hook 'lsp)

  (setq-default c-default-style "linux"
                c-basic-offset 4)

  ;; GDB
  (setq gdb-many-windows t
        gdb-show-main t)

  ;; Helm
  (require 'helm-xref)
  (define-key global-map [remap find-file] #'helm-find-files)
  (define-key global-map [remap execute-extended-command] #'helm-M-x)
  (define-key global-map [remap switch-to-buffer] #'helm-mini)

  ;; Treemacs
  (setq treemacs-space-between-root-nodes nil
        company-idle-delay 0.0
        company-minimum-prefix-length 1
        lsp-idle-delay 0.1)  ;; clangd is fast

  ;; LSP
  (setq lsp-keymap-prefix "C-c l")
  (with-eval-after-load 'lsp-mode
      (add-hook 'lsp-mode-hook #'lsp-enable-which-key-integration)
      (require 'dap-cpptools)
      (yas-global-mode)
      (global-flycheck-mode)
      (lsp-ui-mode))

  (setq lsp-ui-doc-enable t
        lsp-ui-peek-enable t
        lsp-ui-sideline-enable t
        lsp-ui-imenu-enable t
        lsp-ui-flycheck-enable t)


  (require 'fold-this)
  (global-set-key (kbd "C-c C-f") 'fold-this-all)
  (global-set-key (kbd "C-c C-F") 'fold-this)
  (global-set-key (kbd "C-c M-f") 'fold-this-unfold-all)

#+END_SRC

** Haskell

#+BEGIN_SRC emacs-lisp
  (defun haskell-hook ()
    (haskell-indentation-mode)
    (interactive-haskell-mode)
    ;(defvar haskell-font-lock-symbols)
    ;(setq haskell-font-lock-symbols t)
    ;(haskell-unicode-input-method-enable) 
    ;(turn-on-haskell-font-lock)
    )

  (add-hook 'haskell-mode-hook
            (lambda ()
              (flyspell-prog-mode)
              (haskell-hook)))


  ;(add-hook 'haskell-mode-hook 'haskell-hook)
#+END_SRC

** Clojure

#+BEGIN_SRC emacs-lisp      
  (add-hook 'cider-repl-mode-hook #'paredit-mode)
  (add-hook 'cider-mode-hook #'paredit-mode)
  (add-hook 'cider-repl-mode-hook #'company-mode)
  (add-hook 'cider-mode-hook #'company-mode)
  (add-hook 'cider-repl-mode-hook #'cider-company-enable-fuzzy-completion)
  (add-hook 'cider-mode-hook #'cider-company-enable-fuzzy-completion)
#+END_SRC

** R with ESS

#+BEGIN_SRC emacs-lisp
  (use-package ess
    :defer 1
    :config
    (setq inferior-ess-r-program "/usr/bin/R")
    (setq ess-use-company 'script-only)
    (setq ess-use-flymake nil)
    (add-to-list 'auto-mode-alist '("\\.R\\'" . r-mode)))

  (with-eval-after-load 'ess
    (setq ess-use-company t))

  ;; Font highlighting
  (setq ess-R-font-lock-keywords
  '((ess-R-fl-keyword:keywords . t)
    (ess-R-fl-keyword:constants . t)
	(ess-R-fl-keyword:modifiers . t)
	(ess-R-fl-keyword:fun-defs . t)
	(ess-R-fl-keyword:assign-ops . t)
	(ess-R-fl-keyword:%op% . t)
	(ess-fl-keyword:fun-calls . t)
	(ess-fl-keyword:numbers . t)
	(ess-fl-keyword:operators)
	(ess-fl-keyword:delimiters)
	(ess-fl-keyword:=)
	(ess-R-fl-keyword:F&T . t)))

#+END_SRC
** Python

#+BEGIN_SRC emacs-lisp
  (use-package elpy
    :defer 1
    :ensure t
    :init
    (elpy-enable)
    :config 
    (setq elpy-rpc-python-command "python3"))

  (when (load "flycheck" t t)
    (setq elpy-modules (delq 'elpy-module-flymake elpy-modules))
    (add-hook 'elpy-mode-hook 'flycheck-mode))
#+END_SRC

** Lua

#+BEGIN_SRC emacs-lisp
  (use-package lua-mode
    :defer 1
    :ensure t 
    :config 
    (autoload 'lua-mode "lua-mode" "Lua editing mode." t)
    (add-to-list 'auto-mode-alist '("\\.lua$" . lua-mode))
    (add-to-list 'interpreter-mode-alist '("lua" . lua-mode))
#+END_SRC


config.el
(setq user-full-name "Berat Ertural"
      user-mail-address "berat.ertural@rwth-aachen.de")

(setq custom-file "~/.emacs.d/custom.el")
(when (file-exists-p custom-file)
  (load custom-file))

(load-theme 'tango-dark t)

;; Theres a bug in Emacs (Bad Reques)
(setq gnutls-algorithm-priority "NORMAL:-VERS-TLS1.3")

;; Optimization
(setq gc-cons-threshold 100000000 ;  100mb
    read-process-output-max (* 1024 1024)) ; 1 mb

(defalias 'yes-or-no-p 'y-or-n-p)

;; Keep all backup and auto-save files in one directory
(setq backup-directory-alist '(("." . "~/.emacs.d/backups")))
(setq auto-save-file-name-transforms '((".*" "~/.emacs.d/auto-save-list/" t)))

;; UTF-8 encoding
(setq locale-coding-system 'utf-8)
(set-terminal-coding-system 'utf-8)
(set-keyboard-coding-system 'utf-8)
(set-selection-coding-system 'utf-8)
(prefer-coding-system 'utf-8)

(setq-default inhibit-startup-screen t
	      indent-tabs-mode nil
	      indicate-empty-lines t)

(tool-bar-mode -1)
(menu-bar-mode -1)
(show-paren-mode t)
(global-linum-mode t)
(column-number-mode t)
(delete-selection-mode t)

(setq shell-command-switch "-ic")

;; Org-mode settings
(add-to-list 'auto-mode-alist '("\\.org$" . org-mode))
(global-set-key "\C-c l" 'org-store-link)
(global-set-key "\C-c a" 'org-agenda)
(setq org-log-done t)

(require 'org-bullets)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(use-package ob-restclient
  :ensure t)

(org-babel-do-load-languages
 'org-babel-load-languages
 '((python . t)
  (C . t)
  (calc . t)
  (latex . t)
  (java . t)
  (ruby . t)
  (lisp . t)
  (scheme . t)
  (shell . t)
  (sqlite . t)
  (js . t)
  (restclient . t)
  (ledger . t)))

(defun my-org-confirm-babel-evaluate (lang body)
  "Do not confirm evaluation for these languages."
  (not (or (string= lang "C")
	   (string= lang "java")
	   (string= lang "python")
	   (string= lang "emacs-lisp")
	   (string= lang "sqlite"))))
(setq org-confirm-babel-evaluate 'my-org-confirm-babel-evaluate)

(setq ob-async-no-async-languages-alist '("ipython"))

(setq org-src-fontify-natively t
      org-src-window-setup 'current-window
      org-src-strip-leading-and-trailing-blank-lines t
      org-src-preserve-indentation t
      org-src-tab-acts-natively t)

(add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
(add-to-list 'org-structure-template-alist '("py" . "src python"))
(add-to-list 'org-structure-template-alist '("sh" . "src sh"))
(add-to-list 'org-structure-template-alist '("md" . "src markdown"))
(add-to-list 'org-structure-template-alist '("gq" . "src graphql"))

(use-package ox-pandoc
  :no-require t
  :defer 10
  :ensure t)

(use-package which-key
  :config
  (which-key-mode))

(use-package projectile
  :ensure t
  :diminish projectile-mode
  :commands (projectile-mode projectile-switch-project)
  :bind (("C-c p p" . projectile-switch-project)
	 ("C-c p s s" . projectile-ag)
	 ("C-c p s r" . projectile-ripgrep))
  :config
  (setq projectile-keymap-prefix (kbd "C-c p"))
  (projectile-global-mode t)
  (setq projectile-enable-caching t))

(use-package helm
  :disabled t
  :ensure t
  :diminish helm-mode
  :bind (("C-c h" . helm-command-prefix)
	 ("C-x b" . helm-mini)
	 ("C-`" . helm-resume)
	 ("M-x" . helm-M-x)
	 ("C-x C-f" . helm-find-files)
	 ("C-x C-r" . helm-recentf))
  :init
  (require 'helm-config)
  :config
  (setq helm-locate-command "mdfind -interpret -name %s %s"
	helm-ff-newfile-prompt-p nil
	helm-M-x-fuzzy-match t)
  (helm-mode)
  (set-face-attribute 'helm-source-header nil :height 0.1)
  (helm-autoresize-mode 1)
  (setq helm-display-header-line nil)
  (setq helm-autoresize-max-height 30)
  (setq helm-autoresize-min-height 30))

(use-package helm-projectile
  :ensure t
  :after helm-mode
  :commands helm-projectile
  :bind ("C-c p h" . helm-projectile))

(use-package flycheck
  :ensure t
  :defer 10
  :config 
  (setq flycheck-html-tidy-executable "tidy5")
  (global-flycheck-mode)

(use-package yasnippet
  :ensure t
  :diminish yas-minor-mode
  :config
  (add-to-list 'yas-snippet-dirs (locate-user-emacs-file "snippets"))
  (setq yas-indent-line 'fixed)
  (yas-global-mode))

(use-package company
  :ensure t
  :config
  (setq company-tooltip-limit 20)
  (setq company-idle-delay .15)
  (setq company-echo-delay 0)
  (setq company-begin-commands '(self-insert-command))
  (define-key company-active-map (kbd "C-n") #'company-select-next)
  (define-key company-active-map (kbd "C-p") #'company-select-previous))

(use-package undo-tree
  :diminish undo-tree-mode
  :bind (("M-/" . undo-tree-visualize)))

(use-package sr-speedbar
  :ensure t
  :config
  (setq speedbar-show-unknown-files t))

(use-package smartparens
  :ensure smartparens-config
  :diminish smartparens-mode)

;; C/C++ Language Server
(use-package ccls
  :init
  (setq ccls-executable "/home/berat/aur/ccls/Release/ccls")
  :config
  (setq lsp-prefer-flymake nil)
  (setq-default flycheck-disabled-checkers '(c/c++-clang c/c++-cppcheck c/c++-gcc)))

(use-package lsp-mode
  :require ccls
  :hook prog-mode
  :bind (("C-c l" . lsp-keymap-prefix))
  :config
  (setq lsp-clients-clangd-args '("-j=4" "-background-index" "-log=error"))
  (setq lsp-enable-indentation nil)
  (add-hook 'lsp-mode-hook #'lsp-enable-which-key-integration))

;; Company-LSP
(require 'company-lsp)
(push 'company-lsp company-backends)

(use-package tex
  :ensure auctex
  :config
  (setq-default TeX-master nil))

(setq gdb-many-windows t)

(use-package web-mode
  :config
  (add-to-list 'auto-mode-alist '("\\.phtml\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.tpl\\.php\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.html\\.twig\\'" . web-mode))
  (add-to-list 'auto-mode-alist '("\\.html?\\'" . web-mode)))

(use-package slime
  :hook lisp-mode
  :init
  (setq inferior-lisp-program "/usr/bin/sbcl")
  (load "~/quicklisp/clhs-use-local.el" t)
  (load (expand-file-name "~/quicklisp/slime-helper.el")))

(require 'rgbds-mode)
(add-to-list 'auto-mode-alist '("\\.inc\\'" . asm-mode))

(autoload 'lua-mode "lua-mode" "Lua editing mode." t)
(add-to-list 'auto-mode-alist '("\\.lua$" . lua-mode))
(add-to-list 'interpreter-mode-alist '("lua" . lua-mode))
(setq lua-indent-level 2)

(autoload 'glsl-mode "glsl-mode" "GLSL Editing mode." t)
(add-to-list 'auto-mode-alist '("\\.glsl\\'" . glsl-mode))
(add-to-list 'auto-mode-alist '("\\.vert\\'" . glsl-mode))
(add-to-list 'auto-mode-alist '("\\.frag\\'" . glsl-mode))
(add-to-list 'auto-mode-alist '("\\.geom\\'" . glsl-mode))

(defun toggle-text-fold (&optional level)
  "Fold text indented by LEVEL."
  (interactive "P")
  (if (eq selective-display (1+ (current-column)))
      (set-selective-display 0)
    (set-selective-display (or level (1+ (current-column))))))

(global-set-key (kbd "<f2>") #'sr-speedbar-toggle)
(global-set-key (kbd "<f3>") #'toggle-text-fold)

(defun my-c-mode-hook ()
  "C-Style specific properties."
  ;; Default behaviour
  (setq c-basic-offset 4)
  (c-set-offset 'substatement-open 0)  
  (c-set-offset 'case-label 4)

  ;; Company C Headers
  (setq company-c-headers-path-system
	'("/usr/include/c++/9.2.0" 
	  "/usr/avr/include"
	  "/usr/include"
	  "/usr/local/include"))
  (add-to-list 'company-backends 'company-c-headers))

(add-hook 'c-mode-common-hook 'my-c-mode-hook)

(use-package elpy
  :ensure t
  :init
  (elpy-enable)
  :config
  (setq python-shell-interpreter "python3"
	python-shell-interpreter-args "-i"
	elpy-rpc-python-command "python3"))
